File: packages/skill-catalog/thinkwork-admin/scripts/operations/teams.py
"""Team mutations for the thinkwork-admin skill.

Unit 8. `remove_team_agent` and `remove_team_user` are OPT-IN —
server-side `requireAgentAllowsOperation` refuses unless an admin
explicitly added the op name to the agent's
`permissions.operations` jsonb.
"""

from __future__ import annotations

import json
import os
import sys

HERE = os.path.dirname(os.path.abspath(__file__))
PARENT = os.path.normpath(os.path.join(HERE, ".."))
if PARENT not in sys.path:
    sys.path.insert(0, PARENT)

from thinkwork_admin import _begin_mutation, _end_mutation, _graphql, _safe  # noqa: E402

_TEAM_FIELDS = "id name slug description type status budgetMonthlyCents createdAt"
_TEAM_AGENT_FIELDS = "id teamId agentId tenantId role joinedAt"
_TEAM_USER_FIELDS = "id teamId userId tenantId role joinedAt"


@_safe
def create_team(
    tenant_id: str,
    name: str,
    description: str = "",
    type: str = "",  # noqa: A002
    budget_monthly_cents: int = 0,
    idempotency_key: str = "",
) -> str:
    """Create a team in a tenant."""
    ctx = _begin_mutation("create_team")
    input_dict: dict[str, object] = {"tenantId": tenant_id, "name": name}
    if description:
        input_dict["description"] = description
    if type:
        input_dict["type"] = type
    if budget_monthly_cents > 0:
        input_dict["budgetMonthlyCents"] = budget_monthly_cents
    if idempotency_key:
        input_dict["idempotencyKey"] = idempotency_key
    args = {"input": input_dict}
    try:
        result = _graphql(
            f"""mutation($input: CreateTeamInput!) {{
                createTeam(input: $input) {{ {_TEAM_FIELDS} }}
            }}""",
            args,
        )
        _end_mutation(ctx, status="success", arguments=args)
        return json.dumps(result.get("createTeam"))
    except Exception as exc:
        _end_mutation(ctx, status="failed", arguments=args, refusal_reason=type(exc).__name__)
        raise


@_safe
def add_team_agent(
    team_id: str,
    agent_id: str,
    role: str = "member",
    idempotency_key: str = "",
) -> str:
    """Add an agent to a team."""
    ctx = _begin_mutation("add_team_agent")
    input_dict: dict[str, object] = {"agentId": agent_id, "role": role}
    if idempotency_key:
        input_dict["idempotencyKey"] = idempotency_key
    args = {"teamId": team_id, "input": input_dict}
    try:
        result = _graphql(
            f"""mutation($teamId: ID!, $input: AddTeamAgentInput!) {{
                addTeamAgent(teamId: $teamId, input: $input) {{ {_TEAM_AGENT_FIELDS} }}
            }}""",
            args,
        )
        _end_mutation(ctx, status="success", arguments=args)
        return json.dumps(result.get("addTeamAgent"))
    except Exception as exc:
        _end_mutation(ctx, status="failed", arguments=args, refusal_reason=type(exc).__name__)
        raise


@_safe
def add_team_user(
    team_id: str,
    user_id: str,
    role: str = "member",
    idempotency_key: str = "",
) -> str:
    """Add a user to a team."""
    ctx = _begin_mutation("add_team_user")
    input_dict: dict[str, object] = {"userId": user_id, "role": role}
    if idempotency_key:
        input_dict["idempotencyKey"] = idempotency_key
    args = {"teamId": team_id, "input": input_dict}
    try:
        result = _graphql(
            f"""mutation($teamId: ID!, $input: AddTeamUserInput!) {{
                addTeamUser(teamId: $teamId, input: $input) {{ {_TEAM_USER_FIELDS} }}
            }}""",
            args,
        )
        _end_mutation(ctx, status="success", arguments=args)
        return json.dumps(result.get("addTeamUser"))
    except Exception as exc:
        _end_mutation(ctx, status="failed", arguments=args, refusal_reason=type(exc).__name__)
        raise


@_safe
def remove_team_agent(team_id: str, agent_id: str) -> str:
    """Remove an agent from a team. OPT-IN — not default-allowed."""
    ctx = _begin_mutation("remove_team_agent")
    args = {"teamId": team_id, "agentId": agent_id}
    try:
        result = _graphql(
            "mutation($teamId: ID!, $agentId: ID!) { "
            "removeTeamAgent(teamId: $teamId, agentId: $agentId) "
            "}",
            args,
        )
        _end_mutation(ctx, status="success", arguments=args)
        return json.dumps({"removed": bool(result.get("removeTeamAgent"))})
    except Exception as exc:
        _end_mutation(ctx, status="failed", arguments=args, refusal_reason=type(exc).__name__)
        raise


@_safe
def remove_team_user(team_id: str, user_id: str) -> str:
    """Remove a user from a team. OPT-IN — not default-allowed."""
    ctx = _begin_mutation("remove_team_user")
    args = {"teamId": team_id, "userId": user_id}
    try:
        result = _graphql(
            "mutation($teamId: ID!, $userId: ID!) { "
            "removeTeamUser(teamId: $teamId, userId: $userId) "
            "}",
            args,
        )
        _end_mutation(ctx, status="success", arguments=args)
        return json.dumps({"removed": bool(result.get("removeTeamUser"))})
    except Exception as exc:
        _end_mutation(ctx, status="failed", arguments=args, refusal_reason=type(exc).__name__)
        raise


__all__ = [
    "create_team",
    "add_team_agent",
    "add_team_user",
    "remove_team_agent",
    "remove_team_user",
]

File: packages/admin-ops/src/teams.ts
/**
 * Team operations — mirrors
 * packages/skill-catalog/thinkwork-admin/scripts/operations/teams.py.
 * Mutations go through GraphQL (the REST /api/teams handler covers the
 * SPA path but doesn't expose every mutation; parity with the Python
 * skill is easiest via GraphQL).
 */

import type { AdminOpsClient } from "./client.js";
import {
	TEAM_FIELDS,
	TEAM_AGENT_FIELDS,
	TEAM_USER_FIELDS,
} from "./_fields.js";

export interface Team {
	id: string;
	name: string;
	slug: string;
	description: string | null;
	type: string | null;
	status: string;
	budgetMonthlyCents: number | null;
	createdAt: string;
}

export interface TeamAgent {
	id: string;
	teamId: string;
	agentId: string;
	tenantId: string;
	role: string;
	joinedAt: string;
}

export interface TeamUser {
	id: string;
	teamId: string;
	userId: string;
	tenantId: string;
	role: string;
	joinedAt: string;
}

// ---------------------------------------------------------------------------
// Reads
// ---------------------------------------------------------------------------

export async function listTeams(
	client: AdminOpsClient,
	tenantId: string,
): Promise<Team[]> {
	const data = await client.graphql<{ teams: Team[] }>(
		`query($tenantId: ID!) { teams(tenantId: $tenantId) { ${TEAM_FIELDS} } }`,
		{ tenantId },
	);
	return data.teams ?? [];
}

export async function getTeam(client: AdminOpsClient, id: string): Promise<Team | null> {
	const data = await client.graphql<{ team: Team | null }>(
		`query($id: ID!) { team(id: $id) { ${TEAM_FIELDS} } }`,
		{ id },
	);
	return data.team;
}

// ---------------------------------------------------------------------------
// Mutations
// ---------------------------------------------------------------------------

export interface CreateTeamInput {
	tenantId: string;
	name: string;
	description?: string;
	type?: string;
	budgetMonthlyCents?: number;
	idempotencyKey?: string;
}

export async function createTeam(
	client: AdminOpsClient,
	input: CreateTeamInput,
): Promise<Team> {
	const data = await client.graphql<{ createTeam: Team }>(
		`mutation($input: CreateTeamInput!) { createTeam(input: $input) { ${TEAM_FIELDS} } }`,
		{ input },
	);
	return data.createTeam;
}

export interface AddTeamAgentInput {
	agentId: string;
	role?: string;
	idempotencyKey?: string;
}

export async function addTeamAgent(
	client: AdminOpsClient,
	teamId: string,
	input: AddTeamAgentInput,
): Promise<TeamAgent> {
	const data = await client.graphql<{ addTeamAgent: TeamAgent }>(
		`mutation($teamId: ID!, $input: AddTeamAgentInput!) {
			addTeamAgent(teamId: $teamId, input: $input) { ${TEAM_AGENT_FIELDS} }
		}`,
		{ teamId, input },
	);
	return data.addTeamAgent;
}

export interface AddTeamUserInput {
	userId: string;
	role?: string;
	idempotencyKey?: string;
}

export async function addTeamUser(
	client: AdminOpsClient,
	teamId: string,
	input: AddTeamUserInput,
): Promise<TeamUser> {
	const data = await client.graphql<{ addTeamUser: TeamUser }>(
		`mutation($teamId: ID!, $input: AddTeamUserInput!) {
			addTeamUser(teamId: $teamId, input: $input) { ${TEAM_USER_FIELDS} }
		}`,
		{ teamId, input },
	);
	return data.addTeamUser;
}

export async function removeTeamAgent(
	client: AdminOpsClient,
	teamId: string,
	agentId: string,
): Promise<{ removed: boolean }> {
	const data = await client.graphql<{ removeTeamAgent: boolean }>(
		`mutation($teamId: ID!, $agentId: ID!) { removeTeamAgent(teamId: $teamId, agentId: $agentId) }`,
		{ teamId, agentId },
	);
	return { removed: Boolean(data.removeTeamAgent) };
}

export async function removeTeamUser(
	client: AdminOpsClient,
	teamId: string,
	userId: string,
): Promise<{ removed: boolean }> {
	const data = await client.graphql<{ removeTeamUser: boolean }>(
		`mutation($teamId: ID!, $userId: ID!) { removeTeamUser(teamId: $teamId, userId: $userId) }`,
		{ teamId, userId },
	);
	return { removed: Boolean(data.removeTeamUser) };
}
